File: macjet/inspectors/chrome_tab_mapper.py
"""
MacJet — Chrome Tab Mapper
Maps Chrome renderer PIDs to exact tab titles using CDP.

Strategy:
1. `/json` — instant tab list with titles, URLs, and individual WS endpoints
2. `SystemInfo.getProcessInfo` — renderer PIDs with cumulative CPU time
3. Per-tab WS `Performance.getMetrics` — JS heap per tab (fast 1s timeout)
4. Heuristic PID correlation via creation order when exact mapping unavailable
"""
from __future__ import annotations

import asyncio
import json
import urllib.request
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class ChromeTab:
    """A Chrome tab with resource usage info."""
    target_id: str = ""
    title: str = ""
    url: str = ""
    favicon_url: str = ""
    tab_type: str = "page"  # page, background_page, service_worker, etc.
    ws_url: str = ""
    # Resource metrics (filled in asynchronously)
    js_heap_mb: float = 0.0
    task_duration: float = 0.0
    dom_nodes: int = 0
    # PID mapping (best-effort)
    renderer_pid: int = 0
    cpu_time_s: float = 0.0


@dataclass
class ChromeSnapshot:
    """Full Chrome state snapshot."""
    tabs: list[ChromeTab] = field(default_factory=list)
    renderer_pids: dict[int, float] = field(default_factory=dict)  # pid → cpuTime
    total_tabs: int = 0
    total_js_heap_mb: float = 0.0
    has_cdp: bool = False
    cdp_port: int = 0
    error: str = ""


class ChromeTabMapper:
    """Maps Chrome tabs to renderer PIDs via Chrome DevTools Protocol."""

    def __init__(self, cdp_port: int = 9222):
        self._port = cdp_port
        self._latest = ChromeSnapshot()
        self._ws_available = False

    @property
    def latest(self) -> ChromeSnapshot:
        return self._latest

    async def collect(self) -> ChromeSnapshot:
        """Collect Chrome tab data with PID mapping."""
        snapshot = ChromeSnapshot(cdp_port=self._port)

        # Step 1: Get tab list from /json (instant, no WS needed)
        tabs = await self._get_json_tabs()
        if tabs is None:
            snapshot.error = "CDP not available"
            self._latest = snapshot
            return snapshot

        snapshot.has_cdp = True
        snapshot.tabs = tabs
        snapshot.total_tabs = len(tabs)

        # Step 2: Get renderer PIDs from SystemInfo via browser WS
        renderer_pids = await self._get_renderer_pids()
        snapshot.renderer_pids = renderer_pids

        # Step 3: Try per-tab metrics (best-effort, may fail on some tabs)
        await self._enrich_tab_metrics(tabs)

        # Step 4: Correlate PIDs to tabs
        self._correlate_pids(tabs, renderer_pids)

        # Compute totals
        snapshot.total_js_heap_mb = sum(t.js_heap_mb for t in tabs)

        # Sort: by JS heap if we have it, otherwise by cpu_time from PID correlation
        has_heap = any(t.js_heap_mb > 0 for t in tabs)
        if has_heap:
            tabs.sort(key=lambda t: t.js_heap_mb, reverse=True)
        else:
            tabs.sort(key=lambda t: t.cpu_time_s, reverse=True)

        self._latest = snapshot
        return snapshot

    async def _get_json_tabs(self) -> Optional[list[ChromeTab]]:
        """Get tab list from Chrome's /json endpoint."""
        loop = asyncio.get_event_loop()
        try:
            raw = await loop.run_in_executor(None, self._fetch_json, f"http://localhost:{self._port}/json")
            if raw is None:
                return None

            tabs = []
            for entry in raw:
                if entry.get("type") not in ("page", "background_page"):
                    continue
                tabs.append(ChromeTab(
                    target_id=entry.get("id", ""),
                    title=entry.get("title", ""),
                    url=entry.get("url", ""),
                    favicon_url=entry.get("faviconUrl", ""),
                    tab_type=entry.get("type", "page"),
                    ws_url=entry.get("webSocketDebuggerUrl", ""),
                ))
            return tabs
        except Exception:
            return None

    async def _get_renderer_pids(self) -> dict[int, float]:
        """Get renderer PID → cpuTime mapping from SystemInfo.getProcessInfo."""
        try:
            loop = asyncio.get_event_loop()
            version = await loop.run_in_executor(
                None, self._fetch_json, f"http://localhost:{self._port}/json/version"
            )
            if not version:
                return {}

            browser_ws = version.get("webSocketDebuggerUrl", "")
            if not browser_ws:
                return {}

            import websockets
            async with websockets.connect(browser_ws, close_timeout=2, max_size=10*1024*1024) as ws:
                await ws.send(json.dumps({"id": 1, "method": "SystemInfo.getProcessInfo"}))
                resp = json.loads(await asyncio.wait_for(ws.recv(), timeout=5))

                result = {}
                for p in resp.get("result", {}).get("processInfo", []):
                    if p.get("type") == "renderer":
                        result[p["id"]] = p.get("cpuTime", 0)
                return result

        except ImportError:
            return {}
        except Exception:
            return {}

    async def _enrich_tab_metrics(self, tabs: list[ChromeTab]):
        """Get JS heap and DOM node count for each tab via its individual WS endpoint."""
        try:
            import websockets
        except ImportError:
            return

        sem = asyncio.Semaphore(3)  # Max 3 concurrent WS connections

        async def get_metrics(tab: ChromeTab):
            if not tab.ws_url:
                return
            async with sem:
                try:
                    async with websockets.connect(
                        tab.ws_url, close_timeout=1, open_timeout=1.5, max_size=5*1024*1024
                    ) as ws:
                        await ws.send(json.dumps({
                            "id": 1, "method": "Performance.getMetrics"
                        }))
                        resp = json.loads(await asyncio.wait_for(ws.recv(), timeout=1.5))

                        for m in resp.get("result", {}).get("metrics", []):
                            name = m["name"]
                            val = m["value"]
                            if name == "JSHeapUsedSize":
                                tab.js_heap_mb = val / (1024 * 1024)
                            elif name == "TaskDuration":
                                tab.task_duration = val
                            elif name == "Nodes":
                                tab.dom_nodes = int(val)
                except Exception:
                    pass

        await asyncio.gather(*[get_metrics(t) for t in tabs], return_exceptions=True)

    def _correlate_pids(self, tabs: list[ChromeTab], renderer_pids: dict[int, float]):
        """Correlate renderer PIDs to tabs using best-effort heuristics.
        
        Strategy: 
        1. If we have task_duration from per-tab metrics, sort and pair by workload.
        2. If not, assign PIDs to tabs by index (tab order ≈ renderer creation order).
        Not perfect, but gives reasonable attribution for heaviest consumers.
        """
        if not renderer_pids or not tabs:
            return

        # Sort PIDs by cpuTime (heaviest first)
        sorted_pids = sorted(
            renderer_pids.items(),
            key=lambda x: x[1],
            reverse=True,
        )

        # Check if we have per-tab metrics
        tabs_with_duration = [t for t in tabs if t.task_duration > 0]
        
        if tabs_with_duration:
            # Sort tabs by task_duration (heaviest first) and pair
            tabs_with_duration.sort(key=lambda t: t.task_duration, reverse=True)
            for i, tab in enumerate(tabs_with_duration):
                if i < len(sorted_pids):
                    pid, cpu_time = sorted_pids[i]
                    tab.renderer_pid = pid
                    tab.cpu_time_s = cpu_time
        else:
            # Fallback: assign PIDs to tabs by index
            # This is a rough heuristic but at least shows which PIDs are hot
            for i, tab in enumerate(tabs):
                if i < len(sorted_pids):
                    pid, cpu_time = sorted_pids[i]
                    tab.renderer_pid = pid
                    tab.cpu_time_s = cpu_time

    @staticmethod
    def _fetch_json(url: str):
        """Fetch JSON from a URL synchronously."""
        try:
            req = urllib.request.Request(url, headers={"Accept": "application/json"})
            with urllib.request.urlopen(req, timeout=2) as resp:
                return json.loads(resp.read())
        except Exception:
            return None

    def get_tab_for_pid(self, pid: int) -> Optional[ChromeTab]:
        """Get the tab associated with a renderer PID."""
        for tab in self._latest.tabs:
            if tab.renderer_pid == pid:
                return tab
        return None

    def get_domain(self, url: str) -> str:
        """Extract domain from a URL."""
        try:
            from urllib.parse import urlparse
            parsed = urlparse(url)
            return parsed.netloc or parsed.path[:30]
        except Exception:
            return url[:30]

    def format_tab_label(self, tab: ChromeTab, max_len: int = 40) -> str:
        """Format a tab into a readable label for the process tree."""
        title = tab.title or self.get_domain(tab.url) or "untitled"
        if len(title) > max_len:
            title = title[:max_len - 1] + "…"

        parts = [title]
        if tab.js_heap_mb > 10:
            parts.append(f"[{tab.js_heap_mb:.0f}MB JS]")
        if tab.dom_nodes > 1000:
            parts.append(f"[{tab.dom_nodes} nodes]")

        return " ".join(parts)


async def auto_detect_cdp_port() -> int:
    """Auto-detect Chrome's CDP port from its process arguments."""
    import psutil
    for proc in psutil.process_iter(["name", "cmdline"]):
        try:
            name = proc.info.get("name", "")
            if "Google Chrome" not in name and "chrome" not in name.lower():
                continue
            cmdline = proc.info.get("cmdline", []) or []
            for arg in cmdline:
                if arg.startswith("--remote-debugging-port="):
                    port = int(arg.split("=")[1])
                    return port
        except (psutil.NoSuchProcess, psutil.AccessDenied, ValueError):
            continue
    return 0

File: src/inspectors/chrome_tab_mapper.rs
use futures_util::{SinkExt, StreamExt};
use serde_json::Value;
/// MacJet — Chrome Tab Mapper
/// Maps Chrome renderer PIDs to exact tab titles using CDP.
use smol_str::SmolStr;
use std::process::Stdio;
use tokio::process::Command;
use tokio::time::{timeout, Duration};
use tokio_tungstenite::connect_async;
use tokio_tungstenite::tungstenite::protocol::Message;

#[derive(Debug, Clone, Default)]
pub struct ChromeTab {
    pub target_id: SmolStr,
    pub title: SmolStr,
    pub url: SmolStr,
    pub favicon_url: SmolStr,
    pub tab_type: SmolStr,
    pub ws_url: SmolStr,
    pub js_heap_mb: f64,
    pub task_duration: f64,
    pub dom_nodes: u32,
    pub renderer_pid: u32,
    pub cpu_time_s: f64,
}

#[derive(Debug, Clone, Default)]
pub struct ChromeSnapshot {
    pub tabs: Vec<ChromeTab>,
    pub renderer_pids: rustc_hash::FxHashMap<u32, f64>,
    pub total_tabs: u32,
    pub total_js_heap_mb: f64,
    pub has_cdp: bool,
    pub cdp_port: u16,
    pub error: SmolStr,
}

pub struct ChromeTabMapper {
    cdp_port: u16,
    pub latest: ChromeSnapshot,
}

impl Default for ChromeTabMapper {
    fn default() -> Self {
        Self::new(9222)
    }
}

impl ChromeTabMapper {
    pub fn new(cdp_port: u16) -> Self {
        Self {
            cdp_port,
            latest: ChromeSnapshot::default(),
        }
    }

    pub async fn collect(&mut self) -> &ChromeSnapshot {
        let mut snapshot = ChromeSnapshot {
            cdp_port: self.cdp_port,
            ..Default::default()
        };

        let json_tabs_opt = self.get_json_tabs().await;
        if json_tabs_opt.is_none() {
            snapshot.error = SmolStr::new("CDP not available");
            self.latest = snapshot;
            return &self.latest;
        }

        let mut tabs = json_tabs_opt.unwrap();
        snapshot.has_cdp = true;
        snapshot.total_tabs = tabs.len() as u32;

        let renderer_pids = self.get_renderer_pids().await;
        snapshot.renderer_pids = renderer_pids.clone();

        self.enrich_tab_metrics(&mut tabs).await;
        self.correlate_pids(&mut tabs, &renderer_pids);

        snapshot.total_js_heap_mb = tabs.iter().map(|t| t.js_heap_mb).sum();

        let has_heap = tabs.iter().any(|t| t.js_heap_mb > 0.0);
        if has_heap {
            tabs.sort_by(|a, b| {
                b.js_heap_mb
                    .partial_cmp(&a.js_heap_mb)
                    .unwrap_or(std::cmp::Ordering::Equal)
            });
        } else {
            tabs.sort_by(|a, b| {
                b.cpu_time_s
                    .partial_cmp(&a.cpu_time_s)
                    .unwrap_or(std::cmp::Ordering::Equal)
            });
        }

        snapshot.tabs = tabs;
        self.latest = snapshot;
        &self.latest
    }

    async fn get_json_tabs(&self) -> Option<Vec<ChromeTab>> {
        let url = format!("http://localhost:{}/json", self.cdp_port);
        let child = Command::new("curl")
            .args(["-s", "--connect-timeout", "1", &url])
            .stdout(Stdio::piped())
            .stderr(Stdio::null())
            .spawn()
            .ok()?;

        let result = timeout(Duration::from_secs(2), child.wait_with_output())
            .await
            .ok()?
            .ok()?;
        if !result.status.success() || result.stdout.is_empty() {
            return None;
        }

        let arr: Vec<Value> = serde_json::from_slice(&result.stdout).ok()?;
        let mut tabs = Vec::new();

        for entry in arr {
            let ttype = entry.get("type").and_then(|v| v.as_str()).unwrap_or("");
            if ttype != "page" && ttype != "background_page" {
                continue;
            }

            tabs.push(ChromeTab {
                target_id: SmolStr::new(entry.get("id").and_then(|v| v.as_str()).unwrap_or("")),
                title: SmolStr::new(entry.get("title").and_then(|v| v.as_str()).unwrap_or("")),
                url: SmolStr::new(entry.get("url").and_then(|v| v.as_str()).unwrap_or("")),
                favicon_url: SmolStr::new(
                    entry
                        .get("faviconUrl")
                        .and_then(|v| v.as_str())
                        .unwrap_or(""),
                ),
                tab_type: SmolStr::new(ttype),
                ws_url: SmolStr::new(
                    entry
                        .get("webSocketDebuggerUrl")
                        .and_then(|v| v.as_str())
                        .unwrap_or(""),
                ),
                ..Default::default()
            });
        }

        Some(tabs)
    }

    async fn get_renderer_pids(&self) -> rustc_hash::FxHashMap<u32, f64> {
        let mut map = rustc_hash::FxHashMap::default();
        let url = format!("http://localhost:{}/json/version", self.cdp_port);

        let Some(child) = Command::new("curl")
            .args(["-s", "--connect-timeout", "1", &url])
            .stdout(Stdio::piped())
            .stderr(Stdio::null())
            .spawn()
            .ok()
        else {
            return map;
        };

        let Ok(Ok(result)) = timeout(Duration::from_secs(2), child.wait_with_output()).await else {
            return map;
        };
        if !result.status.success() || result.stdout.is_empty() {
            return map;
        }

        let Ok(version) = serde_json::from_slice::<Value>(&result.stdout) else {
            return map;
        };
        let Some(browser_ws) = version.get("webSocketDebuggerUrl").and_then(|v| v.as_str()) else {
            return map;
        };

        if let Ok(Ok((mut ws_stream, _))) =
            timeout(Duration::from_secs(2), connect_async(browser_ws)).await
        {
            let req = serde_json::json!({
                "id": 1,
                "method": "SystemInfo.getProcessInfo"
            });
            let _ = ws_stream.send(Message::Text(req.to_string().into())).await;

            if let Ok(Some(Ok(msg))) = timeout(Duration::from_secs(2), ws_stream.next()).await {
                if let Message::Text(text) = msg {
                    if let Ok(resp) = serde_json::from_str::<Value>(&text) {
                        if let Some(process_infos) = resp
                            .get("result")
                            .and_then(|r| r.get("processInfo"))
                            .and_then(|p| p.as_array())
                        {
                            for p in process_infos {
                                if p.get("type").and_then(|v| v.as_str()) == Some("renderer") {
                                    if let Some(pid) = p.get("id").and_then(|v| v.as_u64()) {
                                        let cpu_time = p
                                            .get("cpuTime")
                                            .and_then(|v| v.as_f64())
                                            .unwrap_or(0.0);
                                        map.insert(pid as u32, cpu_time);
                                    }
                                }
                            }
                        }
                    }
                }
            }
            let _ = ws_stream.close(None).await;
        }

        map
    }

    async fn enrich_tab_metrics(&self, tabs: &mut Vec<ChromeTab>) {
        let mut set = tokio::task::JoinSet::new();
        for (i, tab) in tabs.iter().enumerate() {
            if !tab.ws_url.is_empty() {
                let ws_url = tab.ws_url.to_string();
                set.spawn(async move {
                    if let Ok(Ok((mut ws_stream, _))) =
                        timeout(Duration::from_secs(1), connect_async(&ws_url)).await
                    {
                        let req = serde_json::json!({
                            "id": 1,
                            "method": "Performance.getMetrics"
                        });
                        let _ = ws_stream.send(Message::Text(req.to_string().into())).await;

                        if let Ok(Some(Ok(Message::Text(text)))) =
                            timeout(Duration::from_secs(1), ws_stream.next()).await
                        {
                            if let Ok(resp) = serde_json::from_str::<Value>(text.as_str()) {
                                if let Some(metrics) = resp
                                    .get("result")
                                    .and_then(|r| r.get("metrics"))
                                    .and_then(|m| m.as_array())
                                {
                                    let mut js_heap = 0.0;
                                    let mut dom_nodes = 0;
                                    let mut task_dur = 0.0;
                                    for m in metrics {
                                        let name =
                                            m.get("name").and_then(|v| v.as_str()).unwrap_or("");
                                        let val =
                                            m.get("value").and_then(|v| v.as_f64()).unwrap_or(0.0);
                                        match name {
                                            "JSHeapUsedSize" => js_heap = val / (1024.0 * 1024.0),
                                            "Nodes" => dom_nodes = val as u32,
                                            "TaskDuration" => task_dur = val,
                                            _ => {}
                                        }
                                    }
                                    let _ = ws_stream.close(None).await;
                                    return (i, js_heap, dom_nodes, task_dur);
                                }
                            }
                        }
                        let _ = ws_stream.close(None).await;
                    }
                    (i, 0.0, 0, 0.0)
                });
            } else {
                set.spawn(async move { (i, 0.0, 0, 0.0) });
            }
        }

        while let Some(res) = set.join_next().await {
            if let Ok((i, js, dom, dur)) = res {
                tabs[i].js_heap_mb = js;
                tabs[i].dom_nodes = dom;
                tabs[i].task_duration = dur;
            }
        }
    }

    fn correlate_pids(
        &self,
        tabs: &mut Vec<ChromeTab>,
        renderer_pids: &rustc_hash::FxHashMap<u32, f64>,
    ) {
        if renderer_pids.is_empty() {
            return;
        }

        let mut available_pids: Vec<(u32, f64)> =
            renderer_pids.iter().map(|(&k, &v)| (k, v)).collect();
        // Sort by cpu_time descending as a heuristic
        available_pids.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));

        let mut used_idx = 0;
        for tab in tabs.iter_mut() {
            if used_idx < available_pids.len() {
                tab.renderer_pid = available_pids[used_idx].0;
                tab.cpu_time_s = available_pids[used_idx].1;
                used_idx += 1;
            }
        }
    }
}
